sam/app/template.yaml

```diff
@@ -43,6 +43,13 @@ Parameters:
     Default: LINUX_CONTAINER
     Description: Environment type used by AWS CodeBuild. See the documentation for details (https://docs.aws.amazon.com/codebuild/latest/userguide/create-project.html#create-project-cli).
     Type: String
+  SourceCodeProvider:
+    Type: String
+    Description: Location of your source code repository
+    Default: GitHub
+    AllowedValues:
+      - GitHub
+      - CodeCommit
   GitHubOAuthToken:
```

Comment:
Need to add empty string default values to all GitHub* and CodeCommit* parameters to make them truly optional.

Reply:
hm, not sure I understand this one. The parameters will automatically get an empty string if the user doesn't pass any value to them at launch time. The Rules will then perform the assertions and fail the stack creation if the value is empty.

Do you want the `Default: ''` for consistency with other parameters, or am I missing something here?

Reply:
> The parameters will automatically get an empty string if the user doesn't pass any value to them at launch time.

Are you referring to console behavior? If you use the CloudFormation API, it doesn't do this and you will get an error if you try to create a change set for a template without specifying explicit values for any parameters that do not specify a default value:

```
% cat foo.yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Parameters:
  Param:
    Type: String

Resources:
  Queue:
    Type: 'AWS::SQS::Queue'

% aws cloudformation create-change-set --stack-name foo --change-set-name a1 --change-set-type CREATE --template-body file://foo.yaml

An error occurred (ValidationError) when calling the CreateChangeSet operation: Parameters: [Param] must have values
```

So to truly make the parameters optional, you should add `Default: ''` to them.

Reply:
Ah, good to know. Yes I was referring to how the console behaves. I will add the `Default: ''`  to all CodeCommit* and GitHub* parameters. Thanks for pointing this out.